src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Legacy/LegacySyntaxNodeExtensions.cs

```diff
@@ -258,49 +318,75 @@ public static IEnumerable<SyntaxNode> FlattenSpans(this SyntaxNode node)
         }
     }
 
-    public static SyntaxNode PreviousSpan(this SyntaxNode node)
+    public static SyntaxNode? PreviousSpan(this SyntaxNode node)
     {
-        if (node == null)
+        if (node is null)
         {
             throw new ArgumentNullException(nameof(node));
         }
 
+        var spanData = s_spanDataTable.GetOrCreateValue(node);
+
+        if (spanData.PreviousComputed)
+        {
+            return spanData.Previous;
+        }
+
         var parent = node.Parent;
-        while (parent != null)
+        while (parent is not null)
         {
-            var flattenedSpans = parent.FlattenSpans();
-            var prevSpan = flattenedSpans.LastOrDefault(n => n.EndPosition <= node.Position && n != node);
-            if (prevSpan != null)
+            foreach (var span in parent.FlattenSpansInReverse())
             {
-                return prevSpan;
+                if (span.EndPosition <= node.Position && span != node)
+                {
+                    spanData.Previous = span;
+                    goto Done;
+                }
             }
 
             parent = parent.Parent;
         }
 
-        return null;
+    Done:
+        // Note: if we didn't find a valid span, spanData.Previous will be null.
+        spanData.PreviousComputed = true;
```

Comment:
I'm not entirely certain that this write can't be reordered before the `spanData.Previous = span;` write before the branch, which could result in a different thread observing an invalid `Previous` value. It should probably use a `Volatile.Write`, or perhaps just mark the `PreviousComputed` and `NextComputed` fields as `volatile` to ensure that reads/writes are not ordered before or after them.

Reply:
To simplify, I ended up just locking on `spanData`.